### PR TITLE
handle vcalander documents with multiple VEVENTS not in order

### LINKED ISSUE #23069: vcalander documents with recurrence events that include exemptions or edits, fail to import properly
### Steps to reproduce
1. import [example1.ics.txt](https://github.com/nextcloud/server/files/5289821/example1.ics.txt) or [example2.ics.txt](https://github.com/nextcloud/server/files/5289097/example2.ics.txt)
2. a number of events are missing from the calendar
3. all events appear in the database, however the first/last occurrence fields are incorrect in the  `oc_calendarobjects` table

### Expected behaviour
the first and last occurrence should reflect the earliest and latest dates in the ICS files and all events should be rendered in the calendar UI

### Actual behaviour
because of the way the VCALENDAR is parsed, the first event is given priority and assumed to be the earlier, this leads to a number of incorrect assumptions leading to the objects being stored in `oc_calendarobjects` with the wrong first and/or last occurrence fields

**Nextcloud version:**
20

**Pull Request**
i have a [PR that resolves this issue](https://github.com/nextcloud/server/pull/23070), and am creating this issue to link to the PR for more context

### PR DESCRIPTION
**Pull request by eleith**
when importing ics files from third party services (ex: google), long
running recurrence vcalander objects can contain numerous vevents
relating to past exemptions or edits.

these objects are handled incorrectly, leading to incorrect first and/or
last occurence dates from being inferred. as a result, a number of
imported events are not rendered in the calendar.

this PR makes use of EventIterator's constructor that takes in an array
of VEVENTS to properly handle these complex cases.

Also of note is that other parameter signatures of EventIterator's constructor,
will be deprecated, [according to comments in sabre dav EventIterator.php](https://github.com/nextcloud/3rdparty/blob/master/sabre/vobject/lib/Recur/EventIterator.php#L86)

resolves #23069

### CODE REVIEW

**Comment by eleith** [comment]
i don't see any dav or caldav tests yet, so i have not added any unit tests

**Comment by kesselb** [comment]
> i don't see any dav or caldav tests yet, so i have not added any unit tests

https://www.github.com/nextcloud/server/tree/master/apps%2Fdav%2Ftests%2Funit%2FCalDAV

**Comment by eleith** [comment]
> > i don't see any dav or caldav tests yet, so i have not added any unit tests
> 
> https://www.github.com/nextcloud/server/tree/master/apps%2Fdav%2Ftests%2Funit%2FCalDAV

thanks @kesselb , I'll add tests

**Comment by eleith** [comment]
added two unit tests based off scenarios from importing a large and recent google calendar export that identified these issues

**Review by gjgd — COMMENTED** [review]
Thanks for the PR! I don't know PHP enough to review this so you should ask someone else, but I will note that some lint checks have failed and seem to be blocking.

**Comment by eleith** [comment]
> Thanks for the PR! I don't know PHP enough to review this so you should ask someone else, but I will note that some lint checks have failed and seem to be blocking.

thanks @gjgd, i'm fixing the lint issues and adding another suggested reviewer

**Comment by eleith** [comment]
fyi, i've been using this patch personally for a month now and with it, i'm able to successfully subscribe to a number of external calendars, all of which contain recurring events.

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 1]
```diff
@@ -2382,26 +2383,40 @@ protected function addChange($calendarId, $objectUri, $operation, $calendarType
...
 		$lastOccurrence = null;
 		$uid = null;
 		$classification = self::CLASSIFICATION_PUBLIC;
+		$hasDTSTART = false;
 		foreach ($vObject->getComponents() as $component) {
-			if ($component->name !== 'VTIMEZONE') {
-				$componentType = $component->name;
-				$uid = (string)$component->UID;
-				break;
+			if ($component->name!=='VTIMEZONE') {
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/CalDavBackend.php:2395]:
```suggestion
			if ($component->name !== 'VTIMEZONE') {
```

cc @ChristophWurst seems there is a rule for php-cs-fixer missing.

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 2]
```diff
@@ -2382,26 +2383,40 @@ protected function addChange($calendarId, $objectUri, $operation, $calendarType
...
+				}
 			}
 		}
 		if (!$componentType) {
 			throw new \Sabre\DAV\Exception\BadRequest('Calendar objects must have a VJOURNAL, VEVENT or VTODO component');
 		}
-		if ($componentType === 'VEVENT' && $component->DTSTART) {
-			$firstOccurrence = $component->DTSTART->getDateTime()->getTimeStamp();
+
+		if (count($vEvents) > 0 && $hasDTSTART) {
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/CalDavBackend.php:2414]: If `$hasDTSTART` is true we have at least one element and `count($vEvents)` is superfluous.

Comment (eleith) [reply apps/dav/lib/CalDAV/CalDavBackend.php]: true, if we assume valid events, since dtstart is required.

**Comment by eleith** [comment]
fyi: it took my like five tries to get the builds to pass. saw flakiness on mysql tests, user test and login tests.

**Review by ChristophWurst — APPROVED** [review]

**Review by gjgd — APPROVED** [review]

**Pull request merged**